test/libsolidity/SolidityNatspecJSON.cpp

```diff
@@ -682,6 +682,116 @@ BOOST_AUTO_TEST_CASE(dev_documenting_no_param_description)
 	expectNatspecError(sourceCode);
 }
 
+BOOST_AUTO_TEST_CASE(user_constructor)
+{
+	char const *sourceCode = R"(
+		contract test {
+			/// @notice this is a really nice constructor
+			function test(uint a, uint second) { }
+		}
+	)";
+
+	char const *natspec = R"ABCDEF({
+	"methods" : {
+		"test(uint256,uint256)" : "this is a really nice constructor"
+	}
+	})ABCDEF";
+
+	checkNatspec(sourceCode, natspec, true);
+}
+
+BOOST_AUTO_TEST_CASE(user_constructor_and_function)
+{
+	char const *sourceCode = R"(
+		contract test {
+			/// @notice this is a really nice constructor
+			function test(uint a, uint second) { }
+			/// another multiplier
+			function mul(uint a, uint second) returns(uint d) { return a * 7 + second; }
+		}
+	)";
+
+	char const *natspec = R"ABCDEF({
+	"methods" : {
+		"mul(uint256,uint256)" : {
+			"notice" : "another multiplier"
+		},
+		"test(uint256,uint256)" : "this is a really nice constructor"
+	}
+	})ABCDEF";
+
+	checkNatspec(sourceCode, natspec, true);
+}
+
+BOOST_AUTO_TEST_CASE(dev_constructor)
+{
+	char const *sourceCode = R"(
+		contract test {
+			/// @author Alex
+			/// @param a the parameter a is really nice and very useful
+			/// @param second the second parameter is not very useful, it just provides additional confusion
+			function test(uint a, uint second) { }
+		}
+	)";
+
+	char const *natspec = R"ABCDEF({
+	"methods" : {
+		"test(uint256,uint256)" : {
+			"author" : "Alex",
+			"params" : {
+				"a" : "the parameter a is really nice and very useful",
+				"second" : "the second parameter is not very useful, it just provides additional confusion"
+			}
+		}
+	}
+	})ABCDEF";
+
+	checkNatspec(sourceCode, natspec, false);
+}
+
+BOOST_AUTO_TEST_CASE(dev_constructor_and_function)
+{
+	char const *sourceCode = R"(
+		contract test {
+			/// @author Alex
+			/// @param a the parameter a is really nice and very useful
+			/// @param second the second parameter is not very useful, it just provides additional confusion
+			function test(uint a, uint second) { }
+			/// @dev Multiplies a number by 7 and adds second parameter
+			/// @param a Documentation for the first parameter starts here.
+			/// Since it's a really complicated parameter we need 2 lines
+			/// @param second Documentation for the second parameter
+			/// @return The result of the multiplication
+			/// and cookies with nutella
+			function mul(uint a, uint second) returns(uint d) {
+				return a * 7 + second;
+			}
+		}
+	)";
+
+	char const *natspec = R"ABCDEF({
+	"methods" : {
+		"mul(uint256,uint256)" : {
+			"details" : "Multiplies a number by 7 and adds second parameter",
+			"params" : {
+				"a" : "Documentation for the first parameter starts here. Since it's a really complicated parameter we need 2 lines",
+				"second" : "Documentation for the second parameter"
+			},
+			"return" : "The result of the multiplication and cookies with nutella"
+		},
+		"test(uint256,uint256)" : {
```

Comment:
Not sure if we should use this name. The constructor does not really have a signature.

@axic what do you think?

Reply:
Actually after #3923 it could just use `consturctor(uin256,uin256)` since `constructor` is a restricted name.

Then again, probably natspec needs to be redesigned at this point.

Reply:
@axic ok, i would prefer constructor(..).. the questions is where to place that.. in `methods`, or should we do that within an extra json node...
```
{
"methods" : {
	"mul(uint256,uint256)" : {
		"notice" : "another multiplier"
	},
	"test(uint256,uint256)" : "this is a really nice constructor"
}
}
```
hmm.. interesting.. natspecs on fallbacks.. never tried that...

are there already ideas around "redesigning natspec"?

Reply:
#1141 is the most comprehensive collection of issues

Reply:
@axic oh yeah ;)

Reply:
How about `[constructor](uint256,uint256)`?

@chriseth any preference between the 4 options above?

I think we can merge this once decided and then redesign Natspec anyway.

Reply:
What about `constructor` without parentheses? The key is just used to compute the function signature (note that it does not have names for the arguments). The idea of natspec is that you look at a transaction, check its first four bytes and then find the corresponding function.

Reply:
Sounds good!

Reply:
@aarlt can you please make that change or enable push access for us?

Reply:
@axic @chriseth so the final version should be just `constructor`? I would prefer not to remove potentially useful information.. could imagine that it may be useful for some tools.. am i wrong?

Reply:
however, i think next week i will have some time to checkout the natspec redesign.. anything else that should be done beside stuff mentioned in #1141 and #2243? @chriseth  you wanted that arbitrary natspec tags are allowed, right? additionally to that I would like to be able to distinguish between natspec tags that need to preserve newlines.. or we just introduce a special json node that will hold the tag values with all the newlines.. it would also be nice to keep track of the line number - so it is possible to retrieve the location of a specific natspec tag within the solidity source.. @axic @chriseth what are your thoughts?

Reply:
Can you create a new issue for that please?

Also I think we can just go forward with `constructor` in order to get this finally merged.

Reply:
@axic @chriseth Is it still correct to place `constructor` within the `methods` object?
```json
{
	"methods" : 
	{
		"constructor" : "this is a really nice constructor"
	}
}
```